### PR TITLE
Update autograd docs

### PR DESCRIPTION
**Pull request by zou3519**
There are some FIXMEs right now that I'll make another pass through. Opening this PR for discussion.

### CODE REVIEW

File: torch/autograd/__init__.py [hunk 1]
```diff
@@ -88,32 +86,27 @@ def grad(outputs, inputs, grad_outputs=None, retain_graph=None, create_graph=Fal
...
+            grad. If a None value would be acceptable for all grad_tensors, then this
+            argument is optional.
         retain_graph (bool, optional): If ``False``, the graph used to compute the grad
             will be freed. Note that in nearly all cases setting this option to ``True``
             is not needed and often can be worked around in a much more efficient
             way. Defaults to the value of ``create_graph``.
         create_graph (bool, optional): If ``True``, graph of the derivative will
             be constructed, allowing to compute higher order derivative products.
-            Defaults to ``False``, unless ``grad_variables`` contains at least one
-            non-volatile Variable.
```
Comment (zou3519) [inline torch/autograd/__init__.py]: @colesbury is this still true? I don't think it is but I'm not 100% sure:

I haven't been able to find code otherwise, and tested the following:
```
import torch
import torch.autograd as autograd
x = torch.randn((2, 2), requires_grad=True)
out = x ** 2

# Is a graph for the derivatives created?
grad_output = torch.ones((2, 2), requires_grad=False)
out2 = autograd.grad(out, [x], [grad_output], create_graph=True)
out2[0].backward()  # yes

# Is a graph for the derivatives created?
grad_output = torch.ones((2, 2), requires_grad=True)
out2 = autograd.grad(out, [x], [grad_output], create_graph=False)
out2.requires_grad  # false, so no
```

File: torch/autograd/__init__.py [hunk 2]
```diff
@@ -29,33 +29,31 @@ def _make_grads(outputs, grads):
...
         else:
-            raise TypeError("gradients can be either Tensors, Variables or None, but got " +
+            raise TypeError("gradients can be either Tensors or None, but got " +
                             type(grad).__name__)
     return tuple(new_grads)
 
 
-def backward(variables, grad_variables=None, retain_graph=None, create_graph=False):
-    """Computes the sum of gradients of given variables w.r.t. graph leaves.
+def backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False):
```
Comment (SsnL) [inline torch/autograd/__init__.py]: `grad_variables => grad_tensors` is a breaking change.  Maybe we can have both available and deprecate `grad_variables` first?

Comment (zou3519) [reply torch/autograd/__init__.py]: Yikes... that sounds reasonable to me

File: torch/autograd/__init__.py (hunk 1 above)
Comment (apaszke) [reply torch/autograd/__init__.py]: I think it's correct as you wrote it.

File: torch/autograd/function.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 class _ContextMethodMixin(object):
 
     def save_for_backward(self, *tensors):
+        # FIXME: are saved Tensors now under saved_variables or saved_tensors?
```
Comment (apaszke) [inline torch/autograd/function.py]: Both are equivalent now. We should deprecate `saved_variables`

File: docs/source/autograd.rst [hunk 4]
```diff
@@ -43,27 +30,44 @@ actually lower memory usage by any significant amount. Unless you're operating
...
+    use autograd with tensors. Autograd automatically supports Tensors with
+    ``requires_grad`` set to ``True``. In addition, ``Variable(tensor)`` now returns
+    a :class:`torch.Tensor`. Below please find a quick guide on what has changed:
+
+    - ``Variable(tensor, requires_grad=True)`` is now ``torch.tensor(tensor, requires_grad=True)``
+    - ``var.data`` is now just ``tensor``. If you don't want to record operations on tensors,
+      declare then with ``requires_grad=False`` or detach them from the computation graph with
+      :func:`torch.Tensor.detach`.
+    - Methods such as ``var.backward(), var.detach(), var.register_hook()`` now work on tensors
+      with the same name.
```
Comment (SsnL) [inline docs/source/autograd.rst]: nit: names

File: torch/autograd/__init__.py [hunk 5]
```diff
@@ -88,32 +94,27 @@ def grad(outputs, inputs, grad_outputs=None, retain_graph=None, create_graph=Fal
...
     leaves will still be computed, and will be accumulated into their ``.grad``
     attribute.
 
     Arguments:
-        outputs (sequence of Variable): outputs of the differentiated function.
-        inputs (sequence of Variable): Inputs w.r.t. which the gradient will be
+        outputs (sequence of Tensor): outputs of the differentiated function.
+        inputs (sequence of Tensor): Inputs w.r.t. which the gradient will be
             returned (and not accumulated into ``.grad``).
         grad_outputs (sequence of Tensor): Gradients w.r.t. each output.
```
Comment (SsnL) [inline torch/autograd/__init__.py]: Can you also add `Default: None` here? Thanks!

File: torch/autograd/__init__.py [hunk 6]
```diff
@@ -88,32 +94,27 @@ def grad(outputs, inputs, grad_outputs=None, retain_graph=None, create_graph=Fal
...
-            Any tensors will be automatically converted to Variables that are
-            volatile unless ``create_graph`` is ``True``. None values can be
-            specified for scalar Variables or ones that don't require grad.
-            If a None value would be acceptable for all grad_variables, then
-            this argument is optional.
+            None values can be specified for scalar Tensors or ones that don't require
+            grad. If a None value would be acceptable for all grad_tensors, then this
+            argument is optional.
         retain_graph (bool, optional): If ``False``, the graph used to compute the grad
```
Comment (SsnL) [inline torch/autograd/__init__.py]: Can you also add default value here as well?

File: docs/source/autograd.rst [hunk 7]
```diff
@@ -43,27 +30,44 @@ actually lower memory usage by any significant amount. Unless you're operating
...
-.. autoclass:: Variable
-    :members:
+.. warning::
+    The Variable API has been deprecated: Variables are no longer necessary to
+    use autograd with tensors. Autograd automatically supports Tensors with
+    ``requires_grad`` set to ``True``. In addition, ``Variable(tensor)`` now returns
+    a :class:`torch.Tensor`. Below please find a quick guide on what has changed:
+
+    - ``Variable(tensor, requires_grad=True)`` is now ``torch.tensor(tensor, requires_grad=True)``
+    - ``var.data`` is now just ``tensor``. If you don't want to record operations on tensors,
```
Comment (SsnL) [inline docs/source/autograd.rst]: Could you also a line saying that one can set `requires_grad=True` with tensor factory methods, e.g., `torch.randn(3, 3, requires_grad=True)`?

File: torch/autograd/grad_mode.py [hunk 8]
```diff
@@ -64,12 +64,12 @@ def __exit__(self, *args):
 class set_grad_enabled(object):
     r"""Context-manager that sets gradient calculation to on or off.
 
-    `set_grad_enabled` will enable or disable grads based on its argument `mode`.
+    ``set_grad_enabled`` will enable or disable grads based on its argument ``mode``.
```
Comment (SsnL) [inline torch/autograd/grad_mode.py]: minor nit :``attr:`mode` ``

**Review by ssnl — COMMENTED** [review]
Generally looks great! Have some very minor nits and suggestions. :)

File: torch/autograd/__init__.py (hunk 6 above)
Comment (zou3519) [reply torch/autograd/__init__.py]: The retain_graph default is specified.

**Comment by SsnL** [comment]
Will do another pass tomorrow after CI finishes

File: docs/source/autograd.rst [hunk 9]
```diff
@@ -43,27 +30,52 @@ actually lower memory usage by any significant amount. Unless you're operating
...
 
-.. autoclass:: Variable
-    :members:
+.. warning::
+    The Variable API has been deprecated: Variables are no longer necessary to
+    use autograd with tensors. Autograd automatically supports Tensors with
+    ``requires_grad`` set to ``True``. In addition, ``Variable(tensor)`` now returns
+    a :class:`torch.Tensor`. Below please find a quick guide on what has changed:
+
+    - ``Variable(tensor, requires_grad=True)`` is now ``torch.tensor(tensor, requires_grad=True)``
```
Comment (SsnL) [inline docs/source/autograd.rst]: nit: missing a period

**Review by ssnl — APPROVED** [review]
LGTM

File: docs/source/autograd.rst [hunk 10]
```diff
@@ -43,27 +30,48 @@ actually lower memory usage by any significant amount. Unless you're operating
...
 
-.. autoclass:: Variable
-    :members:
+.. warning::
+    The Variable API has been deprecated: Variables are no longer necessary to
+    use autograd with tensors. Autograd automatically supports Tensors with
+    ``requires_grad`` set to ``True``. In addition, ``Variable(tensor)`` now returns
+    a :class:`torch.Tensor`. Below please find a quick guide on what has changed:
+
+    - ``Variable(tensor, requires_grad=True)`` is now ``torch.tensor(tensor, requires_grad=True)``.
```
Comment (apaszke) [inline docs/source/autograd.rst]: I don’t think we should recommend this. IIRC `tensor` enforces a copy, which is way more expensive than the constructor of Variable. We should have a different way of doing this

File: torch/autograd/__init__.py [hunk 11]
```diff
@@ -64,21 +62,29 @@ def backward(variables, grad_variables=None, retain_graph=None, create_graph=Fal
             be constructed, allowing to compute higher order derivative products.
             Defaults to ``False``.
     """
-    variables = (variables,) if isinstance(variables, Variable) else tuple(variables)
+    if grad_variables is not None:
+        warnings.warn("'grad_variables' is deprecated. Use 'grad_tensors' instead.")
+        if grad_tensors is None:
+            grad_tensors = grad_variables
+        else:
+            warnings.warn("'grad_tensors' and 'grad_variables' (deprecated) arguments both "
```
Comment (apaszke) [inline torch/autograd/__init__.py]: That should be an error. I don’t think you ever want this

File: torch/autograd/__init__.py [hunk 12]
```diff
@@ -64,21 +62,29 @@ def backward(variables, grad_variables=None, retain_graph=None, create_graph=Fal
...
+                          "passed to backward(). Using 'grad_tensors'.")
+
+    tensors = (tensors,) if isinstance(tensors, Variable) else tuple(tensors)
```
Comment (apaszke) [inline torch/autograd/__init__.py]: Should we change this instance check to use Tensor?

File: docs/source/autograd.rst (hunk 10 above)
Comment (zou3519) [reply docs/source/autograd.rst]: I'm not sure if there's a different way of doing this now but I agree it would be nice to have some other way

Comment (colesbury) [reply docs/source/autograd.rst]: In most cases it’s just:

tensor.requires_grad = True

In some cases you want to do:

result = tensor.detach()
result.requires_grad = True

**Comment by lianzhibin** [comment]
module 'torch' has no attribute 'no_grad'

**Comment by SsnL** [comment]
@lianzhibin if you are using 0.3.1 or earlier, then you are looking at the wrong doc. This only updates the doc for master.
